### ISSUE TITLE
Training YOLOv3 for custom dataset

### ISSUE THREAD

**Issue by Aleksei91** [AlexeyAB/darknet#2480]
Good evening, I'm trying to train Yolov3 for my dataset of 50000 images (10 classes). I use general Yolov3 config (for coco), just decreased number of iterations a bit. (110k max_batches, batch size 64). How many iterations I need for training? Is network overfitted already? After 20k iterations mAP stops growing, but error decreasing significantly. Tutorial says that I need 2k iterations per class, is it for batch_size = 64?
I train for 608x608 network with subdivisions = 32 (I hope it is ok). My goal is to analyze high-resolution photos, so I crop them for 832x832 pieces with 25% overlap and pass crops to network. Is it good scenario? 

![chart](https://user-images.githubusercontent.com/36134593/53408732-af671b80-39d8-11e9-8a34-5d8567d54caa.png)
Training time is 1 week on 1080ti.

**Comment by AlexeyAB** [comment]
@Aleksei91 Hi,

> How many iterations I need for training? Is network overfitted already?

No. Just it looks like you have small validation dataset. What is its size?


> My goal is to analyze high-resolution photos, so I crop them for 832x832 pieces with 25% overlap and pass crops to network. Is it good scenario?

Yes.

**Comment by Aleksei91** [comment]
> @Aleksei91 Hi,
> 
> > How many iterations I need for training? Is network overfitted already?
> 
> No. Just it looks like you have small validation dataset. What is its size?
About 20k images, but my dataset (train and validation sets) may have annotation problem, (for example it can have about 5 to 10% mistakes and align of bboxes may not be the best). Is network well suited for this situation? Can it cause wrong result on mAP graph?

**Comment by AlexeyAB** [comment]
> About 20k images, but my dataset (train and validation sets) may have annotation problem, (for example it can have about 5 to 10% mistakes and align of bboxes may not be best). Is network well suited for this situation? Can it cause wrong result on mAP graph?

It affects mAP, but not very much.

* Did you get Initial dataset 120k images and divide it to 2 separate datasets: validation (20%) and training(80%) randomly?

* Are your Validation and Training dataset are very similar, but contain different images?

**Comment by Aleksei91** [comment]
> 
> 
> > About 20k images, but my dataset (train and validation sets) may have annotation problem, (for example it can have about 5 to 10% mistakes and align of bboxes may not be best). Is network well suited for this situation? Can it cause wrong result on mAP graph?
> 
> It affects mAP, but not very much.
> 
>     * Did you get Initial dataset 120k images and divide it to 2 separate datasets: validation (20%) and training(80%) randomly?

Yes. Total dataset size is about 65k, 66% images for training and 33% for validation. 

> 
>     * Are your Validation and Training dataset are very similar, but contain different images?

They may not be similar, but i choose images with random.shuffle in Python. But some classes in my dataset are not well suited for general object detection. It's more like "stuff" objects in coco datasets (pipes, pools, long buildings) so images may not be similar.

**Comment by AlexeyAB** [comment]
> i choose images with **random**.shuffle in Python

It is a good way.

Can you rename your cfg-file to txt-file and attach it?

**Comment by Aleksei91** [comment]
[yolov3-custom.txt](https://github.com/AlexeyAB/darknet/files/2905736/yolov3-custom.txt)

**Comment by AlexeyAB** [comment]
* You can try to train **SPP**-model `yolov3-spp.cfg` file instead of `yolov3.cfg`
* with **default anchors**
* and with pre-trained weights-file `yolov3-spp.conv.85` that you can get by using this command
`./darknet partial cfg/yolov3-spp.cfg yolov3-spp.weights yolov3-spp.conv.85 85`
you should download https://pjreddie.com/media/files/yolov3-spp.weights
* since you get mAP-plato too early, you can slightly reduce `steps=70000,90000`

**Comment by richipower** [comment]
to train yolov3-spp do i need to create yolov3-spp.conv.85 file with the original yolov3-spp.cfg file or with the one modified by me? or it does not matter which one?

**Comment by AlexeyAB** [comment]
@richipower 

> to train yolov3-spp do i need to create yolov3-spp.conv.85 file with the original yolov3-spp.cfg file or with the one modified by me(1 or 2 classes)?

* if you didn't change the first 85 layers in your modifiedyolov3-spp-cfg-file, then you can use it to do `partial`

* if you changed the first 85 layers in your modifiedyolov3-spp-cfg-file, then you **must** use original `yolov3-spp.cfg` to do `partial`

> is for the tiny xnor the same process?

Usually yes, you can use `yolov3-tiny.conv.15` to train yolov3-tiny-xnor

**Comment by cyrineee** [comment]
@Aleksei91
can u share the script that u used to divide the database into train and validation ?
